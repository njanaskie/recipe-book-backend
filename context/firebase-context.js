import React, { useEffect, useState, useContext, useReducer } from 'react';
import database, { firebase, googleAuthProvider } from '../src/firebase/firebase';
import { history } from '../src/routers/AppRouter';
import authReducer from '../src/reducers/auth'
import LoadingPage from '../src/components/LoadingPage'

const FirebaseContext = React.createContext()

export const useFirebaseContext = () => useContext(FirebaseContext)

const FirebaseProvider = ({ children }) => {
    const [auth, authDispatch] = useReducer(authReducer, [])
    // const authDispatch = useDispatch()
    const [user, setUser] = useState('')
    const [loading, setLoading] = useState(true)
    const [isAdmin, setIsAdmin] = useState()
    const [isGuest, setIsGuest] = useState()
    const email = process.env.GUEST_EMAIL
    const password = process.env.GUEST_PASSWORD

    const login = () => {
        firebase.auth().signInWithPopup(googleAuthProvider).then(() => {
            authDispatch({ type: 'LOGIN' })
        })
    }

    const loginAsGuest = () => {
        firebase.auth().signInWithEmailAndPassword(email, password).then(() => {
            authDispatch({ type: 'LOGIN_AS_GUEST' })
        })
    }

    const logout = () => {
        firebase.auth().signOut().then(() => {
            authDispatch({ type: 'LOGOUT' })
        });
    }

    useEffect(() => {
        const unsubscribe = firebase.auth().onAuthStateChanged((user) => {
            setUser(user)
            setLoading(false)
            setIsAdmin()
            setIsGuest()
        })

        return () => unsubscribe()
    }, [])

    useEffect(() => {
        if (user) {
            database.collection('users').doc(user.uid).get()
            .then((doc) => {
                const userData = doc.data()
                if (userData) {
                    if (userData.isAdmin === true) {
                        const isAdmin = userData.isAdmin
                        setIsAdmin(isAdmin)
                    } else if (userData.isGuest === true) {
                        const isGuest = userData.isGuest
                        setIsGuest(isGuest)
                    }
                }
            })
        }
    }, [user])

    if (loading) {
        return <LoadingPage />
    }
    // if (user) {
    //     if (history.location.pathname === '/') {
    //         history.push('/dashboard')
    //     }
    // } else {
    //     history.push('/');
    // }

    return (
        <FirebaseContext.Provider value={{ user, isAdmin, isGuest, login, loginAsGuest, logout }}>
            {children}
        </FirebaseContext.Provider>
    )
}

export { FirebaseProvider, FirebaseContext as default }