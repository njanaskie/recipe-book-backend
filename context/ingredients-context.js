import React from 'react';

const IngredientsContext = React.createContext()

export { IngredientsContext as default }