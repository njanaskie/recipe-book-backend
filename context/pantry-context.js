import React, { useContext } from 'react';

const PantryContext = React.createContext()

export const usePantryContext = () => useContext(PantryContext)

export { PantryContext as default }