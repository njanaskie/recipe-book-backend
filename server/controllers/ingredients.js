const ingredientsRouter = require('express').Router();
const Ingredient = require('../models/ingredient')


getIngredients = async (req, res) => {
    const auth = req.currentUser;

    if (auth) {
        const ingredients = await Ingredient.find({}).sort( { name: 1 } )
        return res.json(ingredients.map((ingredient) => ingredient.toJSON() ));
    }
    return res.status(403).send('Not authorized get')
}
// ingredientsRouter.get('/ingredients', (req, res) => {
//     // const auth = req.currentUser;

//     // if (auth) {
//         // const ingredient = await Ingredient.find({})
//         // console.log('authenticated!', auth);
//         return res.send('Hi, from within the ingredients router GET'); 
//     // }
//     // console.log(req.currentUser)
//     // return res.status(403).send('Not authorized get')
// });

createIngredient = (req, res) => {
    const auth = req.currentUser;

    if (auth) {
        // console.log('authenticated!', auth);
        const ingredient = new Ingredient(req.body)
        const savedIngredient = ingredient.save()

        return res.status(201).json(savedIngredient);
    }
    return res.status(403).send('Not authorized post')

}
// ingredientsRouter.post('/ingredients', (req, res) => {
//     // const auth = req.currentUser;

//     // if (auth) {
//     //     console.log('authenticated!', auth);
//         const ingredient = new Ingredient(req.body)
//         const savedIngredient = ingredient.save()

//         return res.status(201).json(savedIngredient);
//     // }
//     // return res.status(403).send('Not authorized post')
// });

removeIngredient = async (req, res) => {
    const auth = req.currentUser;

    if (auth) {

        await Ingredient.findOneAndDelete({ _id: req.params.id }, (err, ingredient) => {
            if (err) {
                return res.status(400).json({ success: false, error: err })
            }

            if (!ingredient) {
                return res
                    .status(404)
                    .json({ success: false, error: `Ingredient not found` })
            }

            return res.status(200).json({ success: true, data: ingredient })
        }).catch(err => console.log(err))

    }
    return res.status(403).send('Not authorized post')

}

module.exports = {
    getIngredients,
    createIngredient,
    removeIngredient
};