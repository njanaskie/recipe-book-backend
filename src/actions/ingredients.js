import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD_INGREDIENT
export const addIngredient = (ingredient) => ({
    type: 'ADD_INGREDIENT',
    ingredient
});

export const startAddIngredient = (ingredientData = {}) => {
    return (dispatch) => {
        const {
            name = '',
            category = '',
            price = ''
        } = ingredientData;
        const ingredient = { name, category, price };

        return database.ref('ingredients').push(ingredient).then((ref) => {
            dispatch(addIngredient({
                id: ref.key,
                ...ingredient
            }));
        });
    };
};

// REMOVE_INGREDIENT
export const removeIngredient = ({ id } = {}) => ({
    type: 'REMOVE_INGREDIENT',
    id
});

export const startRemoveIngredient = ({ id } = {}) => {
    return (dispatch, getState) => {
        return database.ref(`ingredients/${id}`).remove().then(() => {
            dispatch(removeIngredient({ id }))
        });
    };
};

// EDIT_INGREDIENT
export const editIngredient = (id, updates) => ({
    type: 'EDIT_INGREDIENT',
    id,
    updates
});

export const startEditIngredient = (id, updates) => {
    return (dispatch, getState) => {
        return database.ref(`ingredients/${id}`).update(updates).then(() => {
            dispatch(editIngredient(id, updates))
        });
    }
}

// SET_INGREDIENTS
export const setIngredients = (ingredients) => ({
    type: 'SET_INGREDIENTS',
    ingredients
});

export const startSetIngredients = () => {
    return (dispatch, getState) => {
        return database.ref('ingredients')
        .once('value')
        .then((snapshot) => {
            const ingredients = [];

            snapshot.forEach((childSnapshot) => {
                ingredients.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });

            dispatch(setIngredients(ingredients))
        });
    };
};