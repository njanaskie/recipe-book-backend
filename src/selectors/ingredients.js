
export default (ingredients) => {
    return ingredients.filter(ingredient => ingredient.id !== undefined
    )
    .sort((a, b) => {
        if (a.category === b.category) {
            return b.category - a.category
        }
         return a.category > b.category ? 1 : -1
    })
}