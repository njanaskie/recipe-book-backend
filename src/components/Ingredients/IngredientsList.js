import React, { useContext, useEffect } from 'react'
import _ from 'lodash'
import IngredientListItem from './IngredientListItem'
import IngredientListHeader from './IngredientListHeader'
import IngredientsContext from '../../../context/ingredients-context'
import PantryContext from '../../../context/pantry-context'
import FirebaseContext from '../../../context/firebase-context'
import database from '../../firebase/firebase'
import useIngredients from '../../hooks/useIngredients'
import usePantryIngredients from '../../hooks/usePantryIngredients'
import { useDishesContext } from '../../../context/dishes-context'
import { usePantryDishContext } from '../../../context/pantry-dish-context'
import useDishes from '../../hooks/useDishes'
import useExistingPantryDishes from '../../hooks/useExistingPantryDishes'
import selectPantryDishes from '../../selectors/pantry-dishes'

export const syncIngredientsWithPantry = (ings, pIngs) => {
    const syncedIngredients = []

    if (ings) {
        ings && ings.forEach((ing) => {
            let isPantry = false
            
            pIngs.forEach((pIng) => {
                if (ing.name === pIng.name) {
                    isPantry = true
                } else {
                    isPantry
                }
            })
    
            return syncedIngredients.push({ ...ing, isPantry })
        })
    }

    return syncedIngredients
}

export const IngredientsList = (props) => {
    const ingredients = useIngredients().filter(ingredient => ingredient.id !== undefined)
    const pantryIngredients = usePantryIngredients()
    const syncedIngredients = syncIngredientsWithPantry(ingredients, pantryIngredients)
    const { dishDispatch } = useDishesContext()
    const { pantryDishes, pantryDishDispatch } = usePantryDishContext()
    // const dishes = useDishes()
    // const existingPantryDishes = useExistingPantryDishes()
    // const selectedPantryDishes = selectPantryDishes(pantryIngredients, dishes)

    // console.log(pantryIngredients)
    // console.log(dishes)
    // console.log(existingPantryDishes)
    // console.log(selectedPantryDishes)
    
    const groupedIngredients = _.groupBy(syncedIngredients, 'category')

    return (
        <div className="content-container">
            <div>
                {
                    ingredients ? (
                        Object.keys(groupedIngredients).map(category => {
                            return (
                                <div key={category} >
                                    <IngredientListHeader category={category} />
                                    <div className='ingredient-group-list'>
                                        {groupedIngredients[category].map((ingredient, id) => {
                                            return (
                                                <IngredientListItem
                                                    key={id}
                                                    ingredient={ingredient}
                                                />
                                            )
                                        })}
                                    </div>
                                </div>
                            )
                        })
                    ) : (
                        <div>
                            <span>No ingredients</span>
                        </div>
                    )
                }
            </div>
        </div>
    )
}

export { IngredientsList as default }