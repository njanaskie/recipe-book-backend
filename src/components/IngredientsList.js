import React  from 'react'
import { List } from 'semantic-ui-react'
import IngredientListItem from './IngredientListItem'
import { useIngredientsContext } from '../../context/ingredients-context'
import selectIngredients from '../selectors/ingredients'

export const IngredientsList = () => {
    const { ingredients } = useIngredientsContext()
    const selectedIngredients = selectIngredients(ingredients)

    return (
        <div className="content-container">
            <List divided verticalAlign='middle'>
                {
                    selectedIngredients ? (
                        selectedIngredients.map((ingredient, id) => (
                                <IngredientListItem
                                    key={id}
                                    ingredient={ingredient}
                                />
                        ))
                    ) : (
                        <div>
                            <span>No ingredients</span>
                        </div>
                    )
                }
            </List>
        </div>
    )
}

export { IngredientsList as default }