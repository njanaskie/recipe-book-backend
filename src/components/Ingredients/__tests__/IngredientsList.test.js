import React from 'react'
import { shallow } from 'enzyme'
import { IngredientsList } from '../IngredientsList'
import ingredients from '../../../tests/fixtures/ingredients'
import * as IngredientsContext from '../../../../context/ingredients-context'
import * as FirebaseContext from '../../../../context/firebase-context'
import * as PantryContext from '../../../../context/pantry-context'

let user, pantryIngredients

beforeEach(() => {
    user = jest.spyOn(FirebaseContext, 'useFirebaseContext').mockImplementation(() => user)
    pantryIngredients = jest.spyOn(PantryContext, 'usePantryContext').mockImplementation(() => pantryIngredients)
})

afterEach(() => {
    user.mockRestore()
    pantryIngredients.mockRestore()
})

test('should render IngredientsList with ingredients',() => {
    jest.spyOn(IngredientsContext, 'useIngredientsContext').mockImplementation(() => ingredients)
    const wrapper = shallow(<IngredientsList />)
    expect(wrapper).toMatchSnapshot()
});

// test('should render IngredientsList with ingredients', () => {
//     const wrapper = shallow(
//         <IngredientsContext.Provider value={ingredients}>
//             <IngredientsList ingredients={ingredients}/>
//         </IngredientsContext.Provider >)
//     expect(wrapper).toMatchSnapshot()
// })

// test('should render IngredientsList with empty message', () => {
//     const wrapper = shallow(
//         <IngredientsContext.Provider value={[]}>
//             <IngredientsList ingredients={[]}/>
//         </IngredientsContext.Provider >)
//     expect(wrapper).toMatchSnapshot()
// })