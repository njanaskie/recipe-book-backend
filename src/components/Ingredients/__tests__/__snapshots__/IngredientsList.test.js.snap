// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render IngredientsList with ingredients 1`] = `
<div>
  <div>
    <span>
      No ingredients
    </span>
  </div>
</div>
`;
